### ISSUE TITLE
[Bug] float64 clip not found in opset <= 11

### ISSUE THREAD

**Issue by ganler** [microsoft/onnxruntime#10165]
**Describe the bug**

float64 Clip is supported by ONNX since [opset 6](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#type-constraints-107) and is implemented by ONNXRuntime. But when using ORT to run models on CPU with:

1.  opset <= 11;
2. has a float64 clip node;

It says "[ONNXRuntimeError] : 9 : NOT_IMPLEMENTED : Could not find an implementation for Clip(11) node with name 'Clip_0'". 

But if I separately export the clip node in opset >= 12, it can work. So I think this should be easily fixed by correctly dispatching the operators.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux ise-manjaro 5.10.84-1-MANJARO x64
- ONNX Runtime installed from (source or binary): pip
- ONNX Runtime version: v1.10
- Python version: 3.8

**To Reproduce**

To generate the model:

```python
import torch

class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()

    @torch.no_grad()
    def forward(self, x):
        y = torch.clamp(x, 0, 1)
        return y

model = Model()
inputs = (torch.randn(10,).to(torch.float64),)
torch.onnx.export(model, inputs, "test.onnx", verbose=True,
                  input_names=["input"], 
                  opset_version=11, # if set to >= 12, it works.
                  output_names=["output"])
```

and then use ORT-CPU to run it.

**Expected behavior**

The exported model should be supported by ORT (no errors).

**Screenshots**

export:

![image](https://user-images.githubusercontent.com/38074777/147958909-7dcce335-5da9-49f5-b52f-acfb26012995.png)

run:

![image](https://user-images.githubusercontent.com/38074777/147958927-45b3eb9a-b36d-487b-8710-8760bed74e79.png)

cc: @yuslepukhin

**Comment by yuslepukhin** [comment]
ONNX Runtime has added double (float64) type support to Clip only in opset 12. It is not according to the standard; however, it is not unusual. We sometimes choose not to support some types to save on binary size. I doubt that there would be a minor release to address this.

**Comment by ganler** [comment]
Thanks for your reply.

I see. It makes sense that we don't want to spend extra binary size on old specifications though it can make things incomplete. 

But I do see that it seems some operators can easily achieve spec-complete support w/o extra binary-size overhead. For example (clip opset 11 v.s. 12 in my example):

https://github.com/microsoft/onnxruntime/blob/e9ffcfa24741e4a37dae161b4c81e09b75187156/onnxruntime/core/providers/cpu/math/clip.cc#L20-L27

I am happy to take some time to help investigate more "free" completeness in ORT. :-)

**Comment by ganler** [comment]
compiler should be smart enough to avoid binary size exploding for templated func with the same signature. after enable such kind of data type, binary size does not really grow (only 8 byte diff as is shown in the pic).

![image](https://user-images.githubusercontent.com/38074777/148005512-8bc17d04-889a-4c7e-8f41-eb11c43666bd.png)

**Comment by yuslepukhin** [comment]
Not all of the compilers are equally smart. In general, the amount of code duplication depends on the amount of type dependent code and also the amount of type independent code that is *not* placed in the templated functions. In case of the `Clip`, the general implementation size saves the day although the whole op is unfortunately type dependent. We have applied certain techniques that reduce such code, based on the type size or, if the op is generic enough, such as being reduced to memcpy, we can do it a type independed code.

The overall approach taken for mobile and other size constrained environments is through typelists and configuration files. This allows to create custom builds that support only specifics ops with specific types. Fixed hash values for kernels are a part of this system that produces optimized `ORT` format model files.

You may want to [read up this](https://github.com/microsoft/onnxruntime/blob/master/docs/Reduced_Operator_Kernel_build.md).

Clip implementation features type lists you will want to inspect the usage of that and how that works with the reduced builds and the configuration. This relies on linker's ability on all platforms to remove dead code.

This constrains the changes you can make to previous versions of the product. @edgchen1 that commented on your PR is an expert in the area.

**Comment by skottmckay** [comment]
What's the production use case for a model that uses double and Clip? We generally add support for this sort of thing if there's a real world scenario that requires it. If we implemented support for every type in the onnx spec, a) we'd spend a lot of time doing so and maintaining the code for things that are not required and used in production scenarios, and b) the binary size would be massively larger. 

Unless you absolutely need the accuracy of double, the cost in model size and execution (memory usage and execution time) of using double is significantly higher than float.
